ai_eval/coding_ai_eval.py

```diff
@@ -0,0 +1,252 @@
+            "language": self.language,
+        }
+        frag.initialize_js("CodingAIEvalXBlock", js_data)
+        return frag
+
+    def author_view(self, context=None):
+        """
+        Create preview to be show to course authors in Studio.
+        """
+        if not self.validate():
+            fragment = Fragment()
+            fragment.add_content(
+                _(
+                    "To ensure this component works correctly, please fix the validation issues."
+                )
+            )
+            return fragment
+
+        return self.student_view(context=context)
+
+    def validate_field_data(self, validation, data):
+        """
+        Validate fields
+        """
+
+        super().validate_field_data(validation, data)
+
+        if data.language != LanguageLabels.HTML_CSS and not data.judge0_api_key:
+            validation.add(
+                ValidationMessage(
+                    ValidationMessage.ERROR, _("Judge0 API key is mandatory")
+                )
+            )
+
+    @XBlock.json_handler
+    def get_response(self, data, suffix=""):  # pylint: disable=unused-argument
+        """Get LLM feedback."""
+
+        answer = f"""
+        student code :
+
+        {data['code']}
+        """
+
+        # stdout and stderr only for executable languages (non HTML)
+        if self.language != LanguageLabels.HTML_CSS:
+            answer += f"""
+            stdout:
+
+            {data['stdout']}
+
+            stderr:
+
+            {data['stderr']}
+            """
+
+        messages = [
+            {
+                "role": "system",
+                "content": f"""
+               {self.evaluation_prompt}
+
+               {self.question}.
+
+               The programmimg language is {self.language}
+
+               Evaluation must be in Makrdown format.
+               """,
+            },
+            {
+                "content": f""" Here is the student's answer:
+              {answer}
+                """,
+                "role": "user",
+            },
+        ]
+
+        try:
+            response = get_llm_response(
+                self.model,
+                self.model_api_key,
+                messages,
+                self.model_api_url,
+            )
+
+        except Exception as e:
+            traceback.print_exc()
+            logger.error(
+                f"Failed while making LLM request using model {self.model}. Eaised error type: {type(e)}, Error: {e}"
+            )
+            raise JsonHandlerError(500, "A probem occured. Please retry.") from e
+
+        if response:
+            self.messages[USER_RESPONSE] = data["code"]
+            self.messages[AI_EVALUATION] = response
+            self.messages[CODE_EXEC_RESULT] = {
+                "stdout": data["stdout"],
+                "stderr": data["stderr"],
+            }
+            return {"response": response}
+
+        raise JsonHandlerError(500, "A probem occured. The LLM sent an empty response.")
```

Comment:
The empty response is unexpected, and a 200 status code doesn't seem appropriate. What do you suggest instead of 500?
I changed the message as per your suggestion.

Reply:
I agree that 200 isn't appropriate if an empty response is NOT expected. There are a couple of ways we could go about this:

1. **Explicit:** We return a 500 error to the frontend, with a message like "No AI Feedback was available. Kindly submit again", and show it to the student. Then the student clicks the Submit button again and hopefully the LLM responds back with a valid response.
2. **Implicit:** When the backend detects an empty response, it retries the request for until it gets a non-empty response. Maybe limit the retries to 3 times to avoid infinite retries.

I like the explicit one. We need to display this message in the AI Feedback tab instead of throwing it as an alert, as the error is specific to that one.